### PR TITLE
Moving app from singular to being created for every request

### PR DESCRIPTION
**Pull request by crspeller**
## Summary

This is an MVP of the context refactor. Just enough to get what we want. Additional PRs will follow with some additional cleanup.

EE Changes: https://github.com/mattermost/enterprise/pull/367

## Changes

- The app is used in requests is now created in web/handlers.go ServeHTTP function. Therefore it is created for every request. It has taken some functionality of the context. 
- The context has not gone away, instead the app is a field of the context this allows api4 package scoped fields (like c.Err and c.Params) to exist for now. (This also makes the transition less code change)
- Request scoped related fields are now stored directly on the app this allows them to be used anywhere in the app and passed to plugins (future pr)

## Upcoming improvements

- The main reason for this change to a smaller more focused pr is the complexity in splitting app from server. This involves splitting the server into "services" like the HTTPService. This will be done in future PRs and possibly a community effort.
- Some parts of the app need to be moved to the server as they are services (like config service). Some of these have been left in a state where there are identically named App and Server functions (like config) these will be cleaned up in a future PR.

### CODE REVIEW

File: app/server.go [hunk 1]
```diff
@@ -355,24 +688,7 @@ func (a *App) StartServer() error {
...
-			select {
-			case <-a.Srv.didFinishListen:
-				didShutdown = true
-			case <-timer.C:
-			}
-			timer.Stop()
-		}
-		a.Srv.Server.Close()
-		a.Srv.Server = nil
+	if a.Srv != nil && a.Srv.Server != nil {
```
Comment (jespino) [inline app/server.go]: This looks like no needed code, or something to do.

File: cmd/mattermost/commands/channel_test.go [hunk 2]
```diff
@@ -39,8 +40,12 @@ func TestRemoveChannel(t *testing.T) {
 	// should fail because channel does not exist
 	require.Error(t, RunCommand(t, "channel", "remove", th.BasicTeam.Name+":doesnotexist", th.BasicUser2.Email))
 
+	time.Sleep(time.Second)
```
Comment (jespino) [inline cmd/mattermost/commands/channel_test.go]: If this is needed, something is wrong in the code, or in the test.

File: cmd/mattermost/commands/channel_test.go [hunk 3]
```diff
@@ -39,8 +40,12 @@ func TestRemoveChannel(t *testing.T) {
...
+
 	CheckCommand(t, "channel", "remove", th.BasicTeam.Name+":"+channel.Name, th.BasicUser2.Email)
 
+	time.Sleep(time.Second)
```
Comment (jespino) [inline cmd/mattermost/commands/channel_test.go]: Same here.

File: app/server.go (hunk 1 above)
Comment (crspeller) [reply app/server.go]: Yup. Should be removed!

File: cmd/mattermost/commands/channel_test.go (hunk 2 above)
Comment (crspeller) [reply cmd/mattermost/commands/channel_test.go]: Agreed.

File: api4/integration_action_test.go [hunk 4]
```diff
@@ -84,7 +83,7 @@ func TestOpenDialog(t *testing.T) {
 	pass, resp = Client.OpenInteractiveDialog(request)
 	CheckBadRequestStatus(t, resp)
 	assert.False(t, pass)
-}
```
Comment (jwilander) [inline api4/integration_action_test.go]: This test still causing issues?

Comment (crspeller) [reply api4/integration_action_test.go]: If you pushed a patch with fixes for this I might not have tested it. Will check.

Comment (jwilander) [reply api4/integration_action_test.go]: I didn't push any changes, so it's probably still broken then

**Pull request merged**